java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java

```diff
@@ -266,7 +269,7 @@ public JavaSymbol resolveAs(Tree tree, int kind, Resolve.Env resolveEnv, boolean
         identifierTree = mse.identifier();
         Resolve.Resolution res = getSymbolOfMemberSelectExpression(mse, kind, resolveEnv);
         resolvedSymbol = res.symbol();
-        JavaType resolvedType = resolve.resolveTypeSubstitution(res.type(), getType(mse.expression()));
+        JavaType resolvedType = typeSubstitutionSolver.applySiteSubstitution(res.type(), getType(mse.expression()));
```

Comment:
This is the only place where the typeSubstitutionSolver is used. I tend to think this should be wrapped in Resolve in order to remove the dependendency between the TypeAndRefSolver and the TypeSubstitutionSolver (they will "communicate" through Resolve)